### PR TITLE
move creation of argc/argv symbols into ansi_c_entry_point.cpp

### PR DESCRIPTION
**Pull request by kroening**
The creation of the call for C/C++'s main(argc, argv, envp) is currently
split into two parts: one during typechecking (creates the symbols), and one
afterwards (assigns the symbols).  This commits consolidates both in the
second phase, reducing some complexity.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c9445213b17caa791b46446d0358f68deb368cf5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96402636

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 1]
```diff
@@ -257,7 +257,53 @@ bool generate_ansi_c_start_function(
...
+        argc_symbol.name = "argc'";
+        argc_symbol.type = signed_int_type();
+        argc_symbol.is_static_lifetime = true;
+        argc_symbol.is_lvalue = true;
+        argc_symbol.mode = ID_C;
+
+        if(!symbol_table.insert(std::move(argc_symbol)).second)
+        {
+          messaget message;
+          message.set_message_handler(message_handler);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp]: `messaget message(message_handler);`

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 2]
```diff
@@ -257,7 +257,53 @@ bool generate_ansi_c_start_function(
...
+      }
+
+      const symbolt &argc_symbol = ns.lookup("argc'");
+
+      {
+        // we make the type of this thing an array of pointers
+        // need to add one to the size -- the array is terminated
+        // with NULL
+        const exprt one_expr = from_integer(1, argc_symbol.type);
+        const plus_exprt size_expr(argc_symbol.symbol_expr(), one_expr);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp:285]: `std::move(one_expr)` (and don't make it `const` above)

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 3]
```diff
@@ -257,7 +257,53 @@ bool generate_ansi_c_start_function(
...
+        const array_typet argv_type(pointer_type(char_type()), size_expr);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp:286]: `std::move(size_expr)` (and don't make it `const` above)

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 4]
```diff
@@ -257,7 +257,53 @@ bool generate_ansi_c_start_function(
...
+        argv_symbol.name = "argv'";
+        argv_symbol.type = argv_type;
+        argv_symbol.is_static_lifetime = true;
+        argv_symbol.is_lvalue = true;
+        argv_symbol.mode = ID_C;
+
+        if(!symbol_table.insert(std::move(argv_symbol)).second)
+        {
+          messaget message;
+          message.set_message_handler(message_handler);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp]: `messaget message(message_handler);`

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 5]
```diff
@@ -295,8 +341,44 @@ bool generate_ansi_c_start_function(
...
+          envp_size_symbol.base_name = "envp_size'";
+          envp_size_symbol.name = "envp_size'";
+          envp_size_symbol.type = size_type();
+          envp_size_symbol.is_static_lifetime = true;
+          envp_size_symbol.mode = ID_C;
+
+          if(!symbol_table.insert(std::move(envp_size_symbol)).second)
+          {
+            messaget message;
+            message.set_message_handler(message_handler);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp]: `messaget message(message_handler);`

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 6]
```diff
@@ -295,8 +341,44 @@ bool generate_ansi_c_start_function(
...
+          envp_symbol.name = "envp'";
+          envp_symbol.type = array_typet(
+            pointer_type(char_type()), envp_size_symbol.symbol_expr());
+          envp_symbol.is_static_lifetime = true;
+          envp_symbol.mode = ID_C;
+
+          if(!symbol_table.insert(std::move(envp_symbol)).second)
+          {
+            messaget message;
+            message.set_message_handler(message_handler);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp]: `messaget message(message_handler);` (we should really get rid of the default constructor for `messaget`)

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 7]
```diff
@@ -383,40 +465,36 @@ bool generate_ansi_c_start_function(
...
 
           // disable bounds check on that one
           index_expr.set("bounds_check", false);
 
-          op1=address_of_exprt(index_expr, pointer_type);
+          const pointer_typet &pointer_type =
+            to_pointer_type(parameters[1].type());
+
+          op1 = typecast_exprt::conditional_cast(
+            address_of_exprt(index_expr), pointer_type);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp:478]: `std::move(index_expr)`

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 8]
```diff
@@ -383,40 +465,36 @@ bool generate_ansi_c_start_function(
...
-            from_integer(0, index_type()),
-            pointer_type.subtype());
+            envp_symbol.symbol_expr(), from_integer(0, index_type()));
+
+          const pointer_typet &pointer_type =
+            to_pointer_type(parameters[2].type());
 
-          op2=address_of_exprt(index_expr, pointer_type);
+          operands[2] = typecast_exprt::conditional_cast(
+            address_of_exprt(index_expr), pointer_type);
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_entry_point.cpp:493]: `std::move(index_expr)`

**Review by tautschnig — APPROVED** [review]

**Comment by kroening** [comment]
I've changed the messaget construction, and will make a PR deprecating messaget::messaget().
I am unsure about the copy vs. std::move -- this does not provide a speedup, by may well decrease readability.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3a6fea60e0e7c31868e14c50af89c08d89a18c86).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96443960

**Pull request merged**